rfd/xxxx-tsh-bench-db.md

```diff
@@ -0,0 +1,166 @@
+`tsh bench` for SSH and Kubernetes benchmarking but for database access.
+
+## Why
+The team is working towards understanding performance issues with database
+access. This built-in tool will be used to generate database access load and
+produce some client-side metrics that will be used to understand overall
+performance better.
+
+In addition, the command will also have a mode of direct running the load to the
+databases (without Teleport proxying them). The team can use this to compare
+Teleport performance with direct database access.
+
+## Prior art
+The `tsh` already has commands for performing benchmarks. Those rely on the
+structure defined at `lib/benchmark/benchmark.go`. It defines a small framework
+to define benchmarks. By following it, we can rely on the tools already used on
+the SSH and Kubernetes benchmarks, such as reports.
+
+To help readers better understand how the database benchmark will be organized,
+here is a summary of it:
+* Benchmark suites: Implements a function that generates workload functions.
+  The suite is called before the benchmark starts and returns a workload function. It has access to the Teleport Client.
+* Workload function: Functions that will be repeatedly executed to grab data.
+  Since the suite generates it, it can use information from the suite function.
+  For example, on the SSH benchmark, the servers list used to select the server
+  where the command will be executed is fetched on the suite function, and the
+  workload function only connects to it.
+
+## Details
+The database flow areas will be split into isolated suites, allowing developers
+to test and validate them individually.
+
+The suites will use an interface (`DatabaseClient`) to interact with databases.
+Decoupling it from the suites will increase reusability and also make it quicker
+to enable more database protocols to the benchmark tests.
+
+This interface is returned after connecting to the database. Each protocol will
+define its connection flow respecting the `ConnectFunc`.
+
+```go
+// DatabaseConnectionConfig contains all information necessary to establish a
+// new database connection.
+type DatabaseConnectionConfig struct {
+	// Protocol database protocol.
+	Protocol string
+	// URI direct database connection URI.
+	URI string
+	// Username database username the connection should use.
+	Username string
+	// Database database name where the connection should point to.
+	Database string
+	// ProxyAddress Teleport database proxy address.
+	ProxyAddress string
+	// TLSConfig TLS configuration containing Teleport CA and database
+	// certificates.
+	TLSConfig *tls.Config
+}
+
+// ConnectFunc is a function that establishes a database connection and returns
+// the DatabaseClient.
+type ConnectFunc func(ctx context.Context, config *DatabaseConnectionConfig) (DatabaseClient, error)
+
+// DatabaseClient represents a database connection.
+type DatabaseClient interface {
```

Comment:
I'll place db-specific tests in the future work section. But having a way of providing custom queries ([like Marek mentioned here](https://github.com/gravitational/teleport/pull/27856#discussion_r1230853708)) or even using an external tool (taking the benefit of the [`dbcmd` package](https://github.com/gravitational/teleport/blob/master/lib/client/db/dbcmd/dbcmd.go) to generate the commands) might work better. For example, `tsh bench db external --command="mongosh -f queries.js"`.

My initial thoughts were to use the database clients as little as possible. So we don't have to choose a different driver just because we want to execute complex queries.